hilti/runtime/include/type-info.h

```diff
@@ -1260,6 +1277,10 @@ struct TypeInfo {
     // Actual storage for the held type.
     std::unique_ptr<const char, void (*)(const char*)> _storage = {nullptr, [](const char*) {}};
 
+    // Callback rendering a value of the type into a string. User-code should
+    // use `Value::to_string()` instead of calling this directly.
+    std::string (*const _to_string)(const void* const) = nullptr;
```

Comment:
Just wondering, using a function ptr here is pretty restrictive, e.g., `_to_string` impl cannot be stateful lambda, we always are required to go through a `void*` which forces the nasty `reinterpret_cast` in `TO_STRING`/`TO_STRING_CONST`.

WDYT about instead using a `std::function<std::string(const void* const)>`? That would enable stateful functions and also give us covariance for free, i.e., we could implement functions for more pointers to more specific types, potentially allowing us to avoid `reinterpret_cast` at all (haven't tried this out completely). We'd instantiate one `std::function` per type.